### PR TITLE
ENH: Adding support of GridSearchCV for mne.decoding.LinearModel

### LINKED ISSUE #7027: Adding support of GridSearchCV for mne.decoding.LinearModel
#### Describe the problem
Hi, all~
I recently worked on the decoding analysis using MNE. In my analysis, I used the SVM (linear kernel) and function `SlidingEstimator` to classify my experimental conditions. I also used the `GridSearchCV` (Scikit-learn's function) to find out the best hyper-parameter c of SVM. However, I found that the `LinearModel` of MNE cannot obtain the `coef_` parameter in refitted estimator of `GridSearchCV` when I tried to observe the spatial pattern of refitted estimator.

Here is part of my code:
```python
from sklearn import preprocessing, svm
from sklearn.pipeline import make_pipeline
from sklearn.model_selection import GridSearchCV
from mne.decoding import SlidingEstimator, LinearModel, get_coef

parameters = {'kernel':['linear'], 'C':[1, 10, 100, 1000, 10000]}
# X is a trials*sensors*time matrix.
# y is a vector of condition labels.

clf = make_pipeline(preprocessing.StandardScaler(), LinearModel(GridSearchCV(svm.SVC(), parameters, cv=5, refit=True, iid=False, n_jobs= 5)))
time_decoder = SlidingEstimator(clf, n_jobs=5, scoring='roc_auc', verbose=True)
time_decoder.fit(X, y)
coef = get_coef(time_decoder, 'patterns_', inverse_transform=True) # ValueError
```

#### Describe your solution
From the source code of MNE, we know that the function `get_coef` in `LinearModel` does not consider the refitted estimator of `GridSearchCV`. As we use the `GridSearchCV`, we encounter the ValueError problem because the `coef_` of refitted estimator is located in `time_decoder._final_estimator.best_estimator_.coef_`. Therefore, I was wondering if we can add support of `GridSearchCV` for the class `LinearModel`.

### PR DESCRIPTION
**Pull request by iamsc**
#### Reference issue
Fixes #7027 .

#### What does this implement/fix?
This PR adds the support of `GridSearchCV` for `LinearModel`. I also ran the decoding/tests/test_base.py and got the 3 passed.

### CODE REVIEW

**Comment by agramfort** [comment]
@iamsc can you please add a new test in  https://github.com/mne-tools/mne-python/blob/master/mne/decoding/tests/test_base.py ? thx

**Comment by iamsc** [comment]
> @iamsc can you please add a new test in https://github.com/mne-tools/mne-python/blob/master/mne/decoding/tests/test_base.py ? thx

No problem~ I have modified some tests in `test_base.py`.

Because the original `test_get_coef` function in `test_base.py` only focused on the regression estimator tests, I added some new tests which considers some possible types of estimators, likes classification, regression, classification with GridSearchCV, regression with GridSearchCV.

Moreover, in `test_linearmodel` function, I also added the classification/regression with GridSearchCV tests.

**Comment by agramfort** [comment]
@iamsc you have some flake8 errors. Run `make pep` or flake8 on your files to see what's wrong

**Comment by iamsc** [comment]
> @iamsc you have some flake8 errors. Run `make pep` or flake8 on your files to see what's wrong

Sorry for those errors. I have already fixed those errors on my files using `make pep`.

However, there are still some other errors from `make pep`, likes `mne/tests/test_docstring_parameters.py`. I just ignored this kind of errors because I have not modified those files in this PR.

File: mne/decoding/tests/test_base.py [hunk 1]
```diff
@@ -61,13 +61,32 @@ def test_get_coef():
...
+    assert (is_classifier(lm_classification))
+
+    lm_regression = LinearModel(Ridge())
+    assert (is_regressor(lm_regression))
+
+    parameters = {'kernel': ['linear'], 'C': [1, 10, 100, 1000]}
+    lm_gs_classification = LinearModel(
+        GridSearchCV(
+            svm.SVC(), parameters,
+            cv=5, refit=True, iid=False, n_jobs=5
```
Comment (agramfort) [inline mne/decoding/tests/test_base.py]: for testing try to use something as fast as possible. So cv=2. Also n_jobs should be set to 1 for testing.

File: mne/decoding/tests/test_base.py [hunk 2]
```diff
@@ -61,13 +61,32 @@ def test_get_coef():
...
+        )
+    )
+    assert (is_classifier(lm_gs_classification))
+
+    lm_gs_regression = LinearModel(
+        GridSearchCV(
+            svm.SVR(), parameters, cv=5,
+            refit=True, iid=False, n_jobs=5
```
Comment (agramfort) [inline mne/decoding/tests/test_base.py]: same here

File: mne/decoding/tests/test_base.py [hunk 3]
```diff
@@ -181,10 +232,39 @@ def test_linearmodel():
...
 
-    # check multi-target fit
+    # check categorical target fit in standard linear model with GridSearchCV
+    from sklearn import svm
+    from sklearn.model_selection import GridSearchCV
+    parameters = {'kernel': ['linear'], 'C': [1, 10, 100, 1000]}
+    clf = LinearModel(
+        GridSearchCV(
+            svm.SVC(), parameters, cv=5,
+            refit=True, iid=False, n_jobs=5
```
Comment (agramfort) [inline mne/decoding/tests/test_base.py]: same here

File: mne/decoding/tests/test_base.py [hunk 4]
```diff
@@ -181,10 +232,39 @@ def test_linearmodel():
...
+    assert_equal(clf.patterns_.shape, (n_features,))
+    pytest.raises(ValueError, clf.fit, np.random.rand(n, n_features, 99), y)
+
+    # check continuous target fit in standard linear model with GridSearchCV
+    n_targets = 1
+    Y = np.random.rand(n, n_targets)
+    clf = LinearModel(
+        GridSearchCV(
+            svm.SVR(), parameters, cv=5,
+            refit=True, iid=False, n_jobs=5
```
Comment (agramfort) [inline mne/decoding/tests/test_base.py]: same

**Comment by iamsc** [comment]
> tests should be as fast as possible. Can you report how long it takes to run now compared to before?

If it only sets `cv` as 2, the test duration will decrease 77.17 seconds compared to before.

```
========================================================================== slowest 20 test durations ===========================================================================
180.56s call     mne/decoding/tests/test_base.py::test_get_coef
3.36s setup    mne/decoding/tests/test_base.py::test_get_coef
0.50s call     mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.08s call     mne/decoding/tests/test_base.py::test_linearmodel
0.00s setup    mne/decoding/tests/test_base.py::test_linearmodel
0.00s setup    mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.00s teardown mne/decoding/tests/test_base.py::test_get_coef
0.00s teardown mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.00s teardown mne/decoding/tests/test_base.py::test_linearmodel

Results (184.80s):
       3 passed
```

If I remove 1000 from `C` list in parameter grids and set `cv`  as 2, the test duration will decrease to 15.92 seconds.

```
========================================================================== slowest 20 test durations ===========================================================================
11.61s call     mne/decoding/tests/test_base.py::test_get_coef
3.60s setup    mne/decoding/tests/test_base.py::test_get_coef
0.19s call     mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.03s call     mne/decoding/tests/test_base.py::test_linearmodel
0.00s setup    mne/decoding/tests/test_base.py::test_linearmodel
0.00s setup    mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.00s teardown mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.00s teardown mne/decoding/tests/test_base.py::test_get_coef
0.00s teardown mne/decoding/tests/test_base.py::test_linearmodel

Results (15.69s):
       3 passed
```

I will re-commit the second one later.

**Comment by larsoner** [comment]
Even 11 sec is quite long for a test. How long is it in `master`? Ideally these changes wouldn't add more than 0.1 sec or so. We run a lot of tests, and a lot that do much more complex processing than this, so we need to keep test times to a minimum.

If you need help speeding it up let me know and I can look at tweaking the test to get the time down

**Comment by iamsc** [comment]
> Even 11 sec is quite long for a test. How long is it in `master`? Ideally these changes wouldn't add more than 0.1 sec or so. We run a lot of tests, and a lot that do much more complex processing than this, so we need to keep test times to a minimum.
> 
> If you need help speeding it up let me know and I can look at tweaking the test to get the time down

Here is the time duration results of `master`:
```
========================================================================== slowest 20 test durations ===========================================================================
3.59s setup    mne/decoding/tests/test_base.py::test_get_coef
0.70s call     mne/decoding/tests/test_base.py::test_get_coef
0.23s call     mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.01s call     mne/decoding/tests/test_base.py::test_linearmodel
0.00s setup    mne/decoding/tests/test_base.py::test_linearmodel
0.00s teardown mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.00s teardown mne/decoding/tests/test_base.py::test_get_coef
0.00s setup    mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.00s teardown mne/decoding/tests/test_base.py::test_linearmodel

Results (4.82s):
       3 passed
```

I also tried to remove 100 from `C` list further, the results are as follows:
```
========================================================================== slowest 20 test durations ===========================================================================
4.07s setup    mne/decoding/tests/test_base.py::test_get_coef
1.32s call     mne/decoding/tests/test_base.py::test_get_coef
0.21s call     mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.02s call     mne/decoding/tests/test_base.py::test_linearmodel
0.00s setup    mne/decoding/tests/test_base.py::test_linearmodel
0.00s teardown mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.00s setup    mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.00s teardown mne/decoding/tests/test_base.py::test_linearmodel
0.00s teardown mne/decoding/tests/test_base.py::test_get_coef

Results (6.03s):
       3 passed
```
It seems like to close to the time duration in `master`. I come up with some possible directions I can try with, such as reducing the testing sample size, removing some tests for the possible types of estimators, etc. @larsoner do you have other suggestions that I can also try with?

**Comment by larsoner** [comment]
I would try those, plus reducing the number of samples and/or features.

If the regressors/estimators you are iterating over are `sklearn` estimators, you probably don't need to do more than one. We can trust them to adhere to their own API requirements, so testing a single one should work.

**Comment by iamsc** [comment]
Thanks @larsoner for your suggestions. I have tweaked some tests, including removing some estimators, reducing the parameter space, and reducing the number of samples.

Here is the new time duration:
```
========================================================================== slowest 20 test durations ===========================================================================
3.22s setup    mne/decoding/tests/test_base.py::test_get_coef
0.79s call     mne/decoding/tests/test_base.py::test_get_coef
0.21s call     mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.02s call     mne/decoding/tests/test_base.py::test_linearmodel
0.00s setup    mne/decoding/tests/test_base.py::test_linearmodel
0.00s setup    mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.00s teardown mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.00s teardown mne/decoding/tests/test_base.py::test_linearmodel
0.00s teardown mne/decoding/tests/test_base.py::test_get_coef

Results (4.43s):
       3 passed
```

**Comment by iamsc** [comment]
OK~ No problem.

File: mne/decoding/tests/test_base.py [hunk 5]
```diff
@@ -114,16 +133,43 @@ def inverse_transform(self, X):
...
+    for clf in (lm_regression,
+                lm_gs_classification,
+                make_pipeline(StandardScaler(), lm_classification),
+                make_pipeline(StandardScaler(), lm_gs_regression)):
+
+        # generate some categorical/continuous data
+        # according to the type of estimator.
+        if is_classifier(clf):
+            n, n_features = 1000, 3
+            X = np.random.rand(n, n_features)
```
Comment (agramfort) [inline mne/decoding/tests/test_base.py]: Please use a random state to this test has no effect on global state

File: mne/decoding/tests/test_base.py [hunk 6]
```diff
@@ -181,10 +228,39 @@ def test_linearmodel():
...
+    clf = LinearModel(
+        GridSearchCV(
+            svm.SVC(), parameters, cv=2,
+            refit=True, iid=False, n_jobs=1
+        )
+    )
+    clf.fit(X, y)
+    assert_equal(clf.filters_.shape, (n_features,))
+    assert_equal(clf.patterns_.shape, (n_features,))
+    pytest.raises(ValueError, clf.fit, np.random.rand(n, n_features, 99), y)
```
Comment (agramfort) [inline mne/decoding/tests/test_base.py]: Same here

File: mne/decoding/tests/test_base.py [hunk 7]
```diff
@@ -181,10 +228,39 @@ def test_linearmodel():
...
+
+    # check continuous target fit in standard linear model with GridSearchCV
+    n_targets = 1
+    Y = np.random.rand(n, n_targets)
```
Comment (agramfort) [inline mne/decoding/tests/test_base.py]: Idem

File: mne/decoding/tests/test_base.py [hunk 8]
```diff
@@ -181,10 +228,39 @@ def test_linearmodel():
     assert_equal(clf.patterns_.shape, (n_features,))
     pytest.raises(ValueError, clf.fit, np.random.rand(n, n_features, 99), y)
```
Comment (agramfort) [inline mne/decoding/tests/test_base.py]: Its now recommended to use with block to catch exceptions

File: mne/decoding/tests/test_base.py (hunk 5 above)
Comment (iamsc) [reply mne/decoding/tests/test_base.py]: Sorry for the late replay.
Does this mean that it would be better to use the `np.random.seed(123)` or `np.random.RandomState(123)` before the `np.random.rand()` function?

Comment (larsoner) [reply mne/decoding/tests/test_base.py]: Typically we try to do:
```
def test_whatever():
    rng = np.random.RandomState(0)
    ...
       X = rng.rand(n, n_features)
```

Comment (iamsc) [reply mne/decoding/tests/test_base.py]: Oh! I see~ Let me modify it.

**Comment by agramfort** [comment]
Thx @iamsc !

**Comment by iamsc** [comment]
I’m happy to contribute to MNE and learn from you folks. Thanks @agramfort and @larsoner .

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by agramfort** [comment]
looking forward to your next PRs :)

thanks for contributing

>
